src/core/protocol.ts

```diff
@@ -145,3 +32,81 @@ export class Protocol {
     return request;
   }
 }
+
+export class RedisProtocolError extends Error {
+
+  public static fromMessage(message: string) {
+    const space = message.indexOf(" ");
+    return new RedisProtocolError(message.slice(0, space), message.slice(space + 1));
+  }
+  constructor(name: string, message: string) {
+    super();
+    this.name = name;
+    this.message = message;
+  }
+}
+
+class ProtocolParser {
+  public static parse(raw: string): any[] {
+    const masterRegex = [
+      "(?<simple>(?<=(?:\\r\\n|^)\\+).+?(?=\\r\\n))",
+      "(?<error>(?<=(?:\\r\\n|^)\\-).+?(?=\\r\\n))",
+      "(?<int>(?<=(?:\\r\\n|^)\\:)-?\\d+(?=\\r\\n))",
+      "(?<bulk>(?<=(?:\\r\\n|^)\\$(?<bulk_n>\\d+)\\r\\n).*?(?=\\r\\n))",
+      "(?<null_string>(?<=(?:\\r\\n|^)(\\$|\\*)-1\\r\\n))",
+      "(?<array>(?<=(?:\\r\\n|^)\\*(?<array_n>\\d+)\\r\\n))",
+    ].join("|");
+    return Array.from(raw.matchAll(new RegExp(masterRegex, "g")));
```

Comment:
Absolutely!

So I started here: https://redis.io/topics/protocol, and worked down each of the different data types.

#### RESP Simple Strings
masterRegex[0]: `"(?<simple>(?<=(?:\\r\\n|^)\\+).+?(?=\\r\\n))"`
This statement looks behind (`(?<= )`) for either a newline (`\r\n`) or (`|`) the start of the string (`^`) followed by a literal `+` (`\+`), then captures all characters (`.+?`) up to, but not including the next newline (`(?=\r\n)`).

#### RESP Errors
masterRegex[1]: `"(?<error>(?<=(?:\\r\\n|^)\\-).+?(?=\\r\\n))"`
This statement looks behind for either a newline (`\r\n`) or the start of the string (`^`) followed by a literal `-` (`\-`), then captures all characters (`.+?`) up to, but not including the next newline (`(?=\r\n)`).

#### RESP Integers
masterRegex[2]: `"(?<int>(?<=(?:\\r\\n|^)\\:)-?\\d+(?=\\r\\n))"`
This statement looks behind for either a newline (`\r\n`) or the start of the string (`^`) followed by a literal `:` (`\:`), then optionally captures a negative sign (`-?`) followed by all digits (`\d`) up to, but not including the next newline (`(?=\r\n)`).

#### RESP Bulk Strings
masterRegex[3]: `"(?<bulk>(?<=(?:\\r\\n|^)\\$(?<bulk_n>\\d+)\\r\\n).*?(?=\\r\\n))"`
This statement looks behind for either a newline (`\r\n`) or the start of the string (`^`) followed by a literal `$` (`\$`), then captures the length of the bulk string (`\d+`), followed by a newline, then captures all characters (`.*?`) up to but not including the next newline (`(?=\r\n)`).

#### RESP Null String Special Case
masterRegex[4]: `"(?<null_string>(?<=(?:\\r\\n|^)(\\$|\\*)-1\\r\\n))"`
This statement looks behind for either a newline (`\r\n`) or the start of the string (`^`) followed by either a literal `$-1` or `*-1` (`(\$|\*)-1`), followed by a newline (`\r\n`).

#### RESP Arrays
masterRegex[5]: `"(?<array>(?<=(?:\\r\\n|^)\\*(?<array_n>\\d+)\\r\\n))"`
This statement looks behind for either a newline (`\r\n`) or the start of the string (`^`) followed by a literal `*`, then captures the array length (`\d+`) followed by newline (`\r\n`).  Then outside of the RegExp, we look for N matches of any of these RESP types to fill this array.

This array is then joined with a RexExp _or_ (`|`) which will look for all of these conditions simultaneously.  Admittedly, this regex could probably be simplified a little.  The start condition `(?<=(?:\r\n|^))` is used for each one, so it could probably be factored out. I vaguely remember roughly trying this and didn't get it working, so defaulted to this to start with.

### Aggregation

In base.ts, _on-data_: we append all data to the current string-buffer, then parse the string. 
In protocol.ts, _.parse_: we first match all the above RegExp, then we "aggregate" all the matches according the rules of each RESP type.

> `string.matchAll()` will store any matched named groups in match.groups._name_ where _name_ is defined in the regex like `(?<name> ... )`

1. If we're collecting an Array, `if ("array_n" in current.groups && current.groups.array_n !== undefined)`, we recursively aggregate into the array until all the array elements have been fulfilled.  We only "collect" the array in `output` if we've filled it completely (otherwise we'll try again at the next _on-data_).  The Array itself will be just one element in `output`.
2. If we're collecting a Bulk String, `} else if ("bulk" in current.groups && current.groups.bulk !== undefined) {`, we use a Buffer to check we received the entire Bulk String and if so, "collect" it in `output`.
3. If we're collecting a Null String, `} else if ("null_string" in current.groups && current.groups.null_string !== undefined) {`, we push `null` to `output`.
4. If we're collecting an Error, `} else if ("error" in current.groups && current.groups.error !== undefined) {`, we create an `RedisProtocolError` Object and push to `output`.
5. If we're collecting an Integer, `} else if ("int" in current.groups && current.groups.int !== undefined) {`, we parse the string as an integer and push to `output`.
6. Else, we assume we're collecting a Simple String and just push to `output`

### Flushing

Once the buffer has been parsed and aggregated, we iterated through all collected outputs and callback to each command to flush it from base.

I'm happy to refactor this however you see fit!